actors/migration/nv3/miner.go

```diff
@@ -1,27 +1,426 @@
+func (m *minerMigrator) correctExpirationQueue(exq miner.ExpirationQueue, sectors miner.Sectors,
+	allTerminated bitfield.BitField, allFaults bitfield.BitField, sectorSize abi.SectorSize,
+) (cid.Cid, miner.PowerPair, miner.PowerPair, bitfield.BitField, error) {
+	// processed expired sectors includes all terminated and all sectors seen in earlier expiration sets
+	processedExpiredSectors := allTerminated
+	terminatedAsNonFaults := bitfield.New()
+
+	alteredExpirationSets := make(map[abi.ChainEpoch]miner.ExpirationSet)
+	erasedFaults := bitfield.New()
+	var exs miner.ExpirationSet
+
+	// check for faults that need to be erased
+	err := exq.ForEach(&exs, func(epoch int64) error {
+		// handle faulty sector expirations that have already been processed
+		earlyDuplicates, err := bitfield.IntersectBitField(exs.EarlySectors, processedExpiredSectors)
+		if err != nil {
+			return err
+		} else if empty, err := earlyDuplicates.IsEmpty(); err != nil {
+			return err
+		} else if !empty {
+			alteredExpirationSets[abi.ChainEpoch(epoch)] = exs
+			exs.EarlySectors, err = bitfield.SubtractBitField(exs.EarlySectors, earlyDuplicates)
+			if err != nil {
+				return err
+			}
+
+			terminatedActive, err := bitfield.IntersectBitField(earlyDuplicates, allTerminated)
+			if err != nil {
+				return err
+			}
+			terminatedAsNonFaults, err = bitfield.MergeBitFields(terminatedAsNonFaults, terminatedActive)
+			if err != nil {
+				return err
+			}
+		}
+
+		onTimeDuplicates, err := bitfield.IntersectBitField(exs.OnTimeSectors, processedExpiredSectors)
+		if err != nil {
+			return err
+		} else if empty, err := onTimeDuplicates.IsEmpty(); err != nil {
+			return err
+		} else if !empty {
+			alteredExpirationSets[abi.ChainEpoch(epoch)] = exs
+			exs.OnTimeSectors, err = bitfield.SubtractBitField(exs.OnTimeSectors, onTimeDuplicates)
+			if err != nil {
+				return err
+			}
+
+			terminatedActive, err := bitfield.IntersectBitField(onTimeDuplicates, allTerminated)
+			if err != nil {
+				return err
+			}
+			terminatedAsNonFaults, err = bitfield.MergeBitFields(terminatedAsNonFaults, terminatedActive)
+			if err != nil {
+				return err
+			}
+		}
+
+		if erasedFaults, err = bitfield.MultiMerge(erasedFaults, earlyDuplicates, onTimeDuplicates); err != nil {
+			return err
+		}
+
+		// record on time sectors
+		if processedExpiredSectors, err = bitfield.MultiMerge(processedExpiredSectors, exs.EarlySectors, exs.OnTimeSectors); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	// if we didn't find any faults that needed to be erased, we're done
+	if len(alteredExpirationSets) == 0 {
+		return cid.Undef, miner.PowerPair{}, miner.PowerPair{}, bitfield.BitField{}, nil
+	}
+
+	// save expiration sets that have been altered
+	for epoch, set := range alteredExpirationSets {
+		if err := exq.Set(uint64(epoch), &set); err != nil {
+			return cid.Undef, miner.PowerPair{}, miner.PowerPair{}, bitfield.BitField{}, err
+		}
+	}
```

Comment:
Nit: we could save these during the iteration. This is the safe case of updating an AMT value for a fixed key. It would remove the need to keep track of which ones changed.